trl/trainer/grpo_trainer.py

```diff
@@ -1612,19 +1589,49 @@ def _generate_and_score_completions(
         rewards_per_func = self._calculate_rewards(inputs, original_prompts, completions, completion_ids_list)
 
         # Apply weights to each reward function's output and sum
-        rewards = (rewards_per_func * self.reward_weights.to(device).unsqueeze(0)).nansum(dim=1)
+        scores = (rewards_per_func * self.reward_weights.to(device).unsqueeze(0)).nansum(dim=1)
+
+        # Compute total reward including the KL penalty
+        # Here the KL is computed from the old_policy model to be consistent with the old RLOO/PPO/openRLHF implementation.
+        # However, by definition it has to be computed with respect to the updated_policy! We compute it this way since the 
+        # assumption here is that bc the updated_policy is only ahead by 1 step so it doesn't matter much!
+        if self.beta != 0.0:
+            kl = old_per_token_logps - ref_per_token_logps
+            if self.token_level_kl:
+                # Token-level KL penalty: apply KL penalty per token
+                kl_reward = -self.beta * kl
+
+                # Get the index of the last non-padded token for each sequence
+                eos_indices = (
+                    completion_mask.size(1) - 1 - completion_mask.long().fliplr().argmax(dim=1, keepdim=True)
+                )
+                last_reward = torch.zeros_like(kl)
+                # Ensure scores has correct shape and type
+                scores_shaped = scores.reshape(-1, 1).to(kl.dtype)
+                last_reward.scatter_(dim=1, index=eos_indices, src=scores_shaped)
+
+                # Combine KL reward and last reward
+                reward = last_reward + kl_reward
```

Comment:
Not 100% sure but this line will probably fail in multi gpu setting. 
At this point the rewards are gathered, while the kl is not. You'll need to double check

Reply:
This line or one of the above